### PR TITLE
WIP: Add vboot-tool for vboot2 support

### PR DESCRIPTION
**Pull request by zaolin**

### CODE REVIEW

File: cmds/vboot/vboot.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+)
+
+const (
+	tpmDevice  string = "/dev/tpm0"
+	mountPath  string = "/mnt/vboot"
+	filesystem string = "ext3"
+)
+
+func dieHard() {
+	if e1, e2, err := syscall.Syscall6(syscall.SYS_REBOOT, syscall.LINUX_REBOOT_MAGIC1, syscall.LINUX_REBOOT_MAGIC2, syscall.LINUX_REBOOT_CMD_POWER_OFF, 0, 0, 0); err != 0 {
```
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: just a quick drive-by comment

```go
if err := syscall.Reboot(syscall.LINUX_REBOOT_CMD_POWER_OFF); err != nil {
  log.Fatalf("reboot(CMD_POWER_OFF): %v", err)
}
```

File: cmds/vboot/vboot.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		log.Fatalf("a %v b %v err %v", e1, e2, err)
+	}
+}
+
+func main() {
+	var publicKey = flag.String("pubkey", "/etc/sig.pub", "A public key which should verify the signature.")
```
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: Please declare flags outside of main at the package level, and group them, below the consts:

```go
var (
   pcr = flag.Uint(...
   bootDev = flag.....
)
```

File: cmds/vboot/vboot.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	var linuxKernelSignature = flag.String("kernel-sig", "/mnt/vboot/kernel.sig", "Kernel image signature file path.")
+	var initrd = flag.String("initrd", "/mnt/vboot/initrd", "Initrd file path.")
+	var initrdSignature = flag.String("initrd-sig", "/mnt/vboot/initrd.sig", "Initrd signature file path.")
+	var debug = flag.Bool("debug", false, "Enables debug mode.")
+	var noTPM = flag.Bool("no-tpm", false, "Disables tpm measuring process.")
+
+	flag.Parse()
+
+	if err := os.MkdirAll(mountPath, os.ModePerm); err != nil {
+		if *debug {
```
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: call something like this here or below:

```go
func handleError(err error) {
  if *debug {
    panic(err)
  }
  dieHard()
}
```

File: cmds/vboot/vboot.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
+package main
```
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: Please add a package comment explaining this tool, and briefly explain its usage.

File: cmds/vboot/vboot.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+	if err := os.MkdirAll(mountPath, os.ModePerm); err != nil {
+		if *debug {
+			panic(err)
+		} else {
+			dieHard()
+		}
+	}
+
+	if err := unix.Mount(*bootDev, mountPath, filesystem, unix.MS_RDONLY, ""); err != nil {
```
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: You can just use the syscall package for this, no need for unix. It'd be more consistent to use one or the other, not both.

File: cmds/vboot/vboot.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		if *debug {
+			panic(err)
+		} else {
+			dieHard()
+		}
+	}
+
+	paths := [...]string{*publicKey, *linuxKernel, *linuxKernelSignature, *initrd, *initrdSignature}
```
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: no need for [...], you can just use []string, I believe

File: cmds/vboot/vboot.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+	for _, element := range paths {
+		data, err := ioutil.ReadFile(element)
+		if err != nil {
+			if *debug {
+				panic(err)
+			} else {
+				dieHard()
+			}
+		} else {
```
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: Don't use an else case, when your if always returns or dies. I.e.

```go
data, err := ioutil.ReadFile(element)
if err != nil {
   handleError(err)
}
files[element] = data
```

This serves two purposes: error flows are always indented and hence easily identifiable, but also the expected flow is always indented at the same level.

See 
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: cmds/vboot/vboot.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	kernelDigest := sha256.Sum256(files[*linuxKernel])
+	initrdDigest := sha256.Sum256(files[*initrd])
+
+	pcrDigestKernel := sha1.Sum(files[*linuxKernel])
+	pcrDigestInitrd := sha1.Sum(files[*initrd])
+
+	kernelSuccess := ed25519.Verify(files[*publicKey], kernelDigest[:], files[*linuxKernelSignature])
+	initrdSuccess := ed25519.Verify(files[*publicKey], initrdDigest[:], files[*linuxKernelSignature])
+
+	if kernelSuccess && initrdSuccess {
```
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: Same concept as https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

```go
if !kernelSuccess || !initrdSuccess {
  dieHard()
}

// regular code...
```

File: cmds/vboot/vboot.go (hunk 3 above)
Comment (hugelgupf) [inline cmds/vboot/vboot.go]: Also, https://github.com/golang/go/wiki/CodeReviewComments#dont-panic

log.Errorf may be more appropriate. Or log.Fatalf if you want it to call os.Exit(2).

**Comment by rminnich** [comment]
vboot's mission in life is to do a verified boot. It that's not possible, you can't continue.

File: cmds/vboot/vboot.go (hunk 1 above)
Comment (rminnich) [inline cmds/vboot/vboot.go]: This needs to be in an _linux.go file, e.g. diehard_linux.go. It's very specific to Linux.

File: cmds/vboot/vboot.go (hunk 2 above)
Comment (rminnich) [inline cmds/vboot/vboot.go]: And also look at the other u-root packages as to how to properly do this for busybox mode. see cmds/pwd/pwd.go for an example.

File: cmds/vboot/vboot.go (hunk 5 above)
Comment (rminnich) [inline cmds/vboot/vboot.go]: Not sure I can agree. unix covers darwin, bsd, etc. If you use syscall then the Mount usage should be broken off into e.g. vboot_linux.go. Possibly this mount should be in vboot_unix.go? I realize that u-root has not been consistent about this but we're trying (we originally only envisioned Linux support, but ... well ... you never know).

File: cmds/vboot/vboot.go (hunk 6 above)
Comment (rminnich) [inline cmds/vboot/vboot.go]: Actually I've been dinged on this kind of thing when committing to Go upstream: when I used [] they seem to want [...] since that's an array, not a slice. But who knows.

**Comment by rminnich** [comment]
how's this going? Do you need anything?

**Comment by rminnich** [comment]
phillip, don't give up!

**Comment by hugelgupf** [comment]
Too many comments from me? You need not fix everything, feel free to push back. They're all just suggestions, especially since I don't know anything about vboot. I can be wrong, and I am too nit-picky sometimes!

**Comment by zaolin** [comment]
sorry for the delay. I am too busy atm. I will take care of the requested changes soon this week.
